plugins/outputs/cloudwatchlogs/cloudwatchlogs.go

```diff
@@ -61,17 +61,23 @@ type CloudWatchLogs struct {
 
 	Log telegraf.Logger `toml:"-"`
 
-	cwDests map[Target]*cwDest
+	pusherStopChan  chan struct{}
+	pusherWaitGroup sync.WaitGroup
+	cwDests         map[Target]*cwDest
 }
 
 func (c *CloudWatchLogs) Connect() error {
 	return nil
 }
 
 func (c *CloudWatchLogs) Close() error {
+	close(c.pusherStopChan)
+	c.pusherWaitGroup.Wait()
```

Comment:
Is there a scenario where users would prefer the CWA to stop before sending pending data?
I am wondering about when CloudWatch APIs are timing out.
There is no timeout in this `Close()` function.
So we must rely on on each Pusher to have a reasonable timeout.

Right now I thinking potentially delaying the shutdown, even if it is for a 60 second HTTP timeout, would be preferred by _almost all_ users.

Reply:
It looks to me like `func (p *pusher) send` does not deterministically return in less than 60 seconds because each new  pusher is created like this:
```
pusher := NewPusher(t, client, c.ForceFlushInterval.Duration, maxRetryTimeout, c.Log)
```
where
```
maxRetryTimeout    = 14*24*time.Hour + 10*time.Minute
```
I recommend changing the `send()` function to check the status of:
```
stop                <-chan struct{}
```
at the end of the function before looping and retrying.
So I recommend adding an explicit timeout on

Reply:
> Is there a scenario where users would prefer the CWA to stop before sending pending data?

From a user perspective, I don't personally see any reason to want CWA to drop pending data.  If the user expects/requires CWA to shutdown within a specific timeframe, I would expect them to do so by using the appropriate signals -- namely sending SIGKILL if they actually want the agent to stop immediately..

> Right now I thinking potentially delaying the shutdown, even if it is for a 60 second HTTP timeout, would be preferred by almost all users.

From a practical perspective, most service managers work similar to systemd.  By default systemd will wait 90 seconds after the process is sent the SIGTERM signal, if it still hasn't quit it'll then send a SIGKILL..

Reply:
> I recommend changing the send() function to check the status of:
> ```
> stop                <-chan struct{}
> ```
>
> at the end of the function before looping and retrying.

Addressed this by changing the `time.Sleep(wait)` to a select for either the stop channel or `time.After(wait)`.  This effectively causes all retries to be immediately stopped, which should guarantee that the WaitGroup will return almost immediately.

I added a test to verify this behavior as well.